File: wine_quality_training.py
from pyspark.sql import SparkSession
from pyspark.ml.feature import VectorAssembler
from pyspark.ml.classification import LogisticRegression
from pyspark.ml.evaluation import MulticlassClassificationEvaluator
from pyspark.ml import Pipeline
from pyspark.ml.tuning import ParamGridBuilder, CrossValidator

# Initialize Spark Session
spark = SparkSession.builder \
    .appName("Wine Quality Training") \
    .getOrCreate()

# === Load training dataset ===
training_data = spark.read.format("csv") \
    .option("header", "true") \
    .option("inferSchema", "true") \
    .option("sep", ";") \
    .option("quote", "\"") \
    .load("/app/TrainingDataset.csv")

# === Load validation dataset ===
validation_data = spark.read.format("csv") \
    .option("header", "true") \
    .option("inferSchema", "true") \
    .option("sep", ";") \
    .option("quote", "\"") \
    .load("/app/ValidationDataset.csv")

# === Clean column names ===
def clean_column(col_name):
    return col_name.replace('"', '').replace("'", "").strip()

training_data = training_data.toDF(*[clean_column(c) for c in training_data.columns])
validation_data = validation_data.toDF(*[clean_column(c) for c in validation_data.columns])

# === Extract feature columns ===
feature_columns = training_data.columns
print("Cleaned Columns:", feature_columns)

if "quality" in feature_columns:
    feature_columns.remove("quality")
else:
    raise Exception("Column 'quality' not found even after cleaning.")

# === Create label column and cache ===
training_data = training_data.withColumn("label", training_data["quality"].cast("integer"))
validation_data = validation_data.withColumn("label", validation_data["quality"].cast("integer"))

training_data = training_data.cache()
validation_data = validation_data.cache()

# === Vector Assembler ===
assembler = VectorAssembler(inputCols=feature_columns, outputCol="features")

# === Logistic Regression model ===
lr = LogisticRegression(featuresCol="features", labelCol="label", maxIter=10)

# === Pipeline ===
pipeline = Pipeline(stages=[assembler, lr])

# === Parameter grid for tuning ===
paramGrid = ParamGridBuilder() \
    .addGrid(lr.regParam, [0.01, 0.1, 0.5]) \
    .addGrid(lr.elasticNetParam, [0.0, 0.5, 1.0]) \
    .addGrid(lr.maxIter, [10, 20, 50]) \
    .build()

# === Evaluator (F1 Score) ===
evaluator = MulticlassClassificationEvaluator(
    labelCol="label",
    predictionCol="prediction",
    metricName="f1"
)

# === CrossValidator ===
crossval = CrossValidator(
    estimator=pipeline,
    estimatorParamMaps=paramGrid,
    evaluator=evaluator,
    numFolds=3
)

# === Train model ===
cvModel = crossval.fit(training_data)

# === Get best model ===
best_model = cvModel.bestModel

# === Predict and evaluate on validation data ===
predictions = best_model.transform(validation_data)
f1_score = evaluator.evaluate(predictions)
print(f"\n✅ F1 Score on validation data: {f1_score}\n")

# === Save the best model ===
model_output_path = "/home/jovyan/wine_quality_model"
best_model.write().overwrite().save(model_output_path)
print(f"✅ Model saved to: {model_output_path}")

model_output_path = "/home/jovyan/wine_quality_model"
best_model.write().overwrite().save(model_output_path)
print(f"✅ Model saved to: {model_output_path}")


# === Stop Spark session ===
spark.stop()

File: WineQualityTraining.java
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.ml.feature.VectorAssembler;
import org.apache.spark.ml.classification.LogisticRegression;
import org.apache.spark.ml.evaluation.MulticlassClassificationEvaluator;
import org.apache.spark.ml.Pipeline;
import org.apache.spark.ml.PipelineModel;
import org.apache.spark.ml.tuning.CrossValidator;
import org.apache.spark.ml.tuning.ParamGridBuilder;

public class WineQualityTraining {
    public static void main(String[] args) {
        SparkSession spark = SparkSession.builder()
                .appName("Wine Quality Training")
                .getOrCreate();

        // Load CSVs
        Dataset<Row> trainingData = spark.read().option("header", true)
                .option("inferSchema", true)
                .option("sep", ";")
                .csv("TrainingDataset.csv");

        Dataset<Row> validationData = spark.read().option("header", true)
                .option("inferSchema", true)
                .option("sep", ";")
                .csv("ValidationDataset.csv");

        // Prepare features
        String[] inputCols = trainingData.drop("quality").columns();
        VectorAssembler assembler = new VectorAssembler()
                .setInputCols(inputCols)
                .setOutputCol("features");

        trainingData = trainingData.withColumnRenamed("quality", "label");
        validationData = validationData.withColumnRenamed("quality", "label");

        LogisticRegression lr = new LogisticRegression()
                .setFeaturesCol("features")
                .setLabelCol("label");

        Pipeline pipeline = new Pipeline().setStages(new org.apache.spark.ml.PipelineStage[]{assembler, lr});

        ParamGridBuilder paramGrid = new ParamGridBuilder()
                .addGrid(lr.regParam(), new double[]{0.01, 0.1, 0.5})
                .addGrid(lr.elasticNetParam(), new double[]{0.0, 0.5, 1.0})
                .addGrid(lr.maxIter(), new int[]{10, 20});

        CrossValidator cv = new CrossValidator()
                .setEstimator(pipeline)
                .setEvaluator(new MulticlassClassificationEvaluator().setLabelCol("label").setMetricName("f1"))
                .setEstimatorParamMaps(paramGrid.build())
                .setNumFolds(3);

        PipelineModel bestModel = cv.fit(trainingData);

        double f1 = new MulticlassClassificationEvaluator()
                .setLabelCol("label")
                .setPredictionCol("prediction")
                .setMetricName("f1")
                .evaluate(bestModel.transform(validationData));

        System.out.println("✅ F1 Score: " + f1);

        bestModel.write().overwrite().save("wine_quality_model");

        spark.stop();
    }
}
